### PR TITLE
Add basic Softmax class

### PR DESCRIPTION
**Pull request by anwu1219**
- Add basic SoftmaxConstraint class (currently the only way to solve the a query with Softmax is through preprocessing).
- rename TranscendentalConstraint interface to NonlinearConstraint
- Add supportVariableElimination() method in PLConstraint and NonlinearConstraint (Disjunction and Sigmoid currently does not support variable elimination)
- Change inputOutputVariable in preprocessor

### CODE REVIEW

File: maraboupy/MarabouCore.cpp [hunk 1]
```diff
@@ -97,8 +98,8 @@ void addReluConstraint(InputQuery& ipq, unsigned var1, unsigned var2){
 }
 
 void addSigmoidConstraint(InputQuery& ipq, unsigned var1, unsigned var2){
-    TranscendentalConstraint* s = new SigmoidConstraint(var1, var2);
-    ipq.addTranscendentalConstraint(s);
+    NonlinearConstraint* s = new SigmoidConstraint(var1, var2);
```
Comment (guykatzz) [inline maraboupy/MarabouCore.cpp:101]: * s --> *s

File: maraboupy/MarabouCore.cpp [hunk 2]
```diff
@@ -114,6 +115,18 @@ void addMaxConstraint(InputQuery& ipq, std::set<unsigned> elements, unsigned v){
     ipq.addPiecewiseLinearConstraint(m);
 }
 
+void addSoftmaxConstraint(InputQuery& ipq, std::list<unsigned> inputs,
```
Comment (guykatzz) [inline maraboupy/MarabouCore.cpp:118]: Use List instead of std::list?

Comment (guykatzz) [reply maraboupy/MarabouCore.cpp:118]: Also, add spaces to the parameter lists
void addSoftmaxConstraint( InputQuery& ipq, std::list<unsigned> inputs,
                                             ^

File: src/engine/Engine.h [hunk 3]
```diff
@@ -279,7 +279,7 @@ class Engine : public IEngine, public SignalHandler::Signalable
     /*
       The existing transcendental constraints.
     */
-    List<TranscendentalConstraint *> _tsConstraints;
+    List<NonlinearConstraint *> _tsConstraints;
```
Comment (guykatzz) [inline src/engine/Engine.h:282]: _tsConstraints should probably be renamed, too

File: src/engine/SoftmaxConstraint.cpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        tightenings.append( Tightening( _outputs[i], 1, Tightening::UB ) );
+
+        // First get tighter lower bound
+        bool validBound = true;
+        Vector<double> values;
+        for ( unsigned j = 0; j < _inputs.size(); ++j )
+        {
+            if ( i == j )
+            {
+                if ( existsLowerBound( _inputs[j] ) &&
```
Comment (guykatzz) [inline src/engine/SoftmaxConstraint.cpp:188]: I didn't understand what was going on here. Are we just taking the existing bounds and propagating them as is?

File: maraboupy/MarabouCore.cpp (hunk 2 above)
Comment (anwu1219) [reply maraboupy/MarabouCore.cpp:118]: Since we are going to call this function in python using pybind, I think we have to use std::list<unsigned>

Comment (anwu1219) [reply maraboupy/MarabouCore.cpp:118]: The format in MarabouCore.cpp is completely different from the rest of the code base btw. Probably need a clang pass to fix those.

File: src/engine/Engine.h (hunk 3 above)
Comment (anwu1219) [reply src/engine/Engine.h:282]: done

**Pull request closed without merging**